Prevent disconnected Node Variants when moving a document, in a setup with dimensions

Pingu501 opened the pull request:
ToDo:
- [x] testcases for fallback dimensions
- [x] testcases for dimensions without fallbacks
- [x] implementation of the fix

Resolves: #3384

kdambekalns commented on Neos.ContentRepository/Classes/Domain/Service/NodeMoveIntegrityCheckService.php:
```diff
@@ -0,0 +1,161 @@
+ *
+ * Now, the user moves /site/cr/subpage underneath /site/other/ in the user workspace.
+ *
+ * Until October 2021, this lead to the following situation / bug:
```
```suggestion
 * Until this was merged, this lead to the following situation / bug:
```

Otherwise this will eventually confuse the hell out of someone in the future…

kdambekalns reviewed (commented):
Just a quick feedback after reading the changes… see inline comment and add some more type declarations to properties in the new code. Anything that is not subject to (lazy) injection can be typed…

Also, why does this target only `master`? This is a bugfix, so it should go into the lowest maintained branch. Or is it "impossible" to fix in lower branches?

Pingu501 commented:
@kdambekalns I'm not sure what you mean by `more type declarations to properties`. Can you give me a file and line number and what is missing for example?

It does only target master because I did not had maintenance for older versions in mind. So you think this should be merged into the branch of version 4.3? As far as I'm aware this should be compatible with those older versions.

kdambekalns commented:
> @kdambekalns I'm not sure what you mean by `more type declarations to properties`. Can you give me a file and line number and what is missing for example?

Two properties in `NodeMoveIntegrityCheckResult`, e.g.

    protected string $parts;

**But ignore me**, that is supported as of PHP 7.4 only, so impossible for us, due to https://docs.neos.io/cms/installation-development-setup/system-requirements#php-version-matrix

> It does only target master because I did not had maintenance for older versions in mind. So you think this should be merged into the branch of version 4.3? As far as I'm aware this should be compatible with those older versions.

Well, Neos 5.3 and up is still being supported, so that would be the version to target, if possible.

Pingu501 commented:
@kabarakh after having a detailed look into the code I'm sure the bugfix will be compatible with Neos 5.3, so I rebased my branch to the `5.3` branch and switched the base for this PR to `5.3`. Is this the correct way?

skurfuerst reviewed (approved):
great work and awesome testcases 👍 ❤️
